AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/InitialSync/InitialSyncOrchestrator.swift

```diff
@@ -85,26 +97,31 @@ final class AWSInitialSyncOrchestrator: InitialSyncOrchestrator {
 
     /// Enqueues sync operations for models and downstream dependencies
     private func enqueueSyncOperation(for modelType: Model.Type) {
-        let syncOperationCompletion: SyncOperationResultHandler = { result in
-            if case .failure(let dataStoreError) = result {
-                let syncError = DataStoreError.sync(
-                    "An error occurred syncing \(modelType.modelName)",
-                    "",
-                    dataStoreError)
-                self.syncErrors.append(syncError)
-            }
-        }
-
         let initialSyncForModel = InitialSyncOperation(modelType: modelType,
                                                        api: api,
                                                        reconciliationQueue: reconciliationQueue,
                                                        storageAdapter: storageAdapter,
-                                                       dataStoreConfiguration: dataStoreConfiguration,
-                                                       completion: syncOperationCompletion)
+                                                       dataStoreConfiguration: dataStoreConfiguration)
+
+        initialSyncOperationSinks[modelType.modelName] = initialSyncForModel
+            .publisher
+            .sink(receiveCompletion: { result in
+                if case .failure(let dataStoreError) = result {
+                    let syncError = DataStoreError.sync(
+                        "An error occurred syncing \(modelType.modelName)",
+                        "",
+                        dataStoreError)
+                    self.syncErrors.append(syncError)
+                }
+            }, receiveValue: onReceiveValue(receiveValue:))
 
         syncOperationQueue.addOperation(initialSyncForModel)
     }
 
+    private func onReceiveValue(receiveValue: InitialSyncOperationEvent) {
+        initialSyncOrchestratorTopic.send(receiveValue)
+    }
+
     private func resolveCompletion() {
```

Comment:
Refactor this so that we can re-use the logic for the publisher as well.
```suggestion
    private func onReceiveCompletion() {
        guard initialSyncOperationSinks.isEmpty else {
            return
        }
        completionResult = makeCompletionResult()
        switch {
            case .success:
                initialSyncOrchestratorTopic.send(completion: .finished)
            case .failure(let error):
                initialSyncOrchestratorTopic.send(completion: .failure(error))
        }
    }
    
    private func resolveCompletionBlock() {
        completion?(makeCompletionResult())
    }
    
    private func makeCompletionResult() -> Result<Void, DataStoreError> {
        guard syncErrors.isEmpty else {
            let allMessages = syncErrors.map { String(describing: $0) }
            let syncError = DataStoreError.sync(
                "One or more errors occurred syncing models. See below for detailed error description.",
                allMessages.joined(separator: "\n")
            )
            return .failure(syncError)
        }
        return .successfulVoid
    }
```